### ISSUE TITLE
Why collector always collects two more episodes than I want

### ISSUE THREAD

**Issue by Cra2yDavid** [aai-institute/tianshou#555]
- [ ] I have marked all applicable categories:
    + [x] exception-raising bug
    + [x] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [ ] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [ ] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, numpy, sys
  print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```
I want to test all samples in my test set in every test. Howevery I notice the collector always collects two more episodes. For example, there are 5 samples in test env. And I rewrote method RESET(evaluation=True) to: sample_idx = (sample_idx+1)%5 and PRINT sample_idx (0~4). And I set the parameter 'episode_per_test' to 5. However, the collector always collects 7 episodes in test, which shows on console is '0, 1, 2, 3, 4, 0, 1' and in the next test is '2, 3, 4, 0, 1, 2, 3'. While it should work as '0, 1, 2, 3, 4', '0, 1, 2, 3, 4' and so on. How can I deal with it. Were the two extra episodes used in test?

**Comment by Trinkle23897** [comment]
Could you please paste a simple script here for reproducing the buggy result?

**Comment by Cra2yDavid** [comment]
something like this, may also work in Cartpole env:

```
def __init__(self, args, evaluation=False):
        self.sample_idx = 0
        self.test_sample_num = 5
def reset(self):
        self.current_step = 0
        if self.evaluation:
            self.sample_idx = (self.sample_idx+1) % self.test_sample_num
            print(self.sample_idx, end=',')
```

**Comment by Trinkle23897** [comment]
```python
class Wrapper(gym.Wrapper):
    def __init__(self, env, s):
        self.num = s
        self.env = env
        self.cnt = 0

    def reset(self):
        print(self.num, self.cnt)
        self.cnt = 0
        return super().reset()

    def step(self, act):
        self.cnt += 1
        return super().step(act)

train_envs = DummyVectorEnv(
    [lambda x=_: Wrapper(gym.make(args.task), s=f"train env_id={x}, len=") for _ in range(args.training_num)]
)
test_envs = DummyVectorEnv(
    [lambda x=_: Wrapper(gym.make(args.task), s=f"test env_id={x}, len=") for _ in range(args.test_num)]
)
```

Running `python3 test/discrete/test_dqn.py --test-num 5`, the output is

```
test env_id=0, len= 0
test env_id=1, len= 0
test env_id=2, len= 0
test env_id=3, len= 0
test env_id=4, len= 0
test env_id=0, len= 200
test env_id=1, len= 200
test env_id=2, len= 200
test env_id=3, len= 200
test env_id=4, len= 200
test env_id=0, len= 0
test env_id=1, len= 0
test env_id=2, len= 0
test env_id=3, len= 0
test env_id=4, len= 0
```

So I think it's reasonable, right? One environment runs exactly one episode. However, the number of "useless" reset can be reduced...

**Comment by Cra2yDavid** [comment]
Or to be specific, I think the Collector didn't run as I want. I have applied some changes on your code like:
```
args.test_num = 1

class Wrapper(gym.Wrapper):
    def __init__(self, env, s):
        self.num = s
        self.env = env
        self.cnt = 0

    def reset(self):
        print(self.num, self.cnt)
        self.cnt += 1  # each time the RESET() is called, cnt+=1
        return super().reset()

test_envs = DummyVectorEnv(
    [lambda x=_: Wrapper(gym.make(args.task), s=f"test env_id={x}, episode=") for _ in range(args.test_num)]
)

net = Net(args.state_shape, args.action_shape, hidden_sizes=args.hidden_sizes, device=args.device).to(args.device)

optim = torch.optim.Adam(net.parameters(), lr=args.lr)

policy = DQNPolicy(net, optim, args.gamma, args.n_step)

test_collector = Collector(policy, test_envs)
test_collector.collect(n_episode=5)  # the collector should collect 5 episodes and call RESET() method for 5 times. 
```
So I want to collect 5 episodes in one single env. The output is:

> test env_id=0, episode= 0
> test env_id=0, episode= 1
> test env_id=0, episode= 2
> test env_id=0, episode= 3
> test env_id=0, episode= 4
> test env_id=0, episode= 5
> test env_id=0, episode= 6
> 
The reset() method was called for 7 times, which should be 5 as I set. It puzzuled me a lot.

**Comment by Trinkle23897** [comment]
I think it should be at least 6, because collector cannot assume the vector_env doesn't have unfinished episode (what if some user manually call `collector.collect(n_step)` and then `collector.collect(n_episode)`?). The first and last reset are useless because of https://github.com/thu-ml/tianshou/blob/d976a5aa913218634e62957f35621f8f43f38f77/tianshou/data/collector.py#L113 and https://github.com/thu-ml/tianshou/blob/d976a5aa913218634e62957f35621f8f43f38f77/tianshou/data/collector.py#L323 Maybe we can remove the last one?

**Comment by Cra2yDavid** [comment]
Thanks for your answer
